metafeed-finder.js

```diff
@@ -0,0 +1,229 @@
+const pull = require('pull-stream')
+const debug = require('debug')('ssb:replication-scheduler')
+const detectSsbNetworkErrorSeverity = require('ssb-network-errors')
+const { where, type, live, toPullStream } = require('ssb-db2/operators')
+const { validateMetafeedAnnounce } = require('ssb-meta-feeds/validate')
+
+const DEFAULT_PERIOD = 500
+
+module.exports = class MetafeedFinder {
+  constructor(ssb, opts, period) {
+    this._ssb = ssb
+    this._opts = opts
+    this._period = period || DEFAULT_PERIOD
+    this._map = new Map() // mainFeedId => rootMetaFeedId
+    this._inverseMap = new Map() // rootMetaFeedId => mainFeedId
+    this._requestsByMainfeedId = new Map() // mainFeedId => Array<Calback>
+    this._latestRequestTime = 0
+    this._timer = null
+
+    // If at least one hops template is configured, then load
+    if (
+      this._opts.partialReplication &&
+      Object.values(this._opts.partialReplication).some((templ) => !!templ)
+    ) {
+      if (!this._ssb.db || !this._ssb.db.query) {
+        throw new Error(
+          'ssb-replication-scheduler expects ssb-db2 to be installed, to use partial replication'
+        )
+      }
+      this._loadAllFromLog()
+    }
+  }
+
+  fetch(mainFeedId, cb) {
+    if (this._map.has(mainFeedId)) {
+      const metaFeedId = this._map.get(mainFeedId)
+      cb(null, metaFeedId)
+      return
```

Comment:
these returns in this function is interfering with my ocd brain. Can we remove this one. I'm ok with the return cb(err) further down, but that one can go as well.